File: dashboard-scripts/fetchData.php
<?php
header("Access-Control-Allow-Origin: *");
header("Content-Type: application/json");

$config = require __DIR__ . '/config/DB/dbConfig.php';

// Database connection details
$servername = $config['servername'];
$username = $config['username'];
$password = $config['password'];
$dbname = $config['dbname'];

// Set the default timezone to your local timezone
date_default_timezone_set('America/New_York');

try {
    // Create a new PDO instance with persistent connection
    // Set the PDO error mode to exception
    $pdo = new PDO("mysql:host=$servername;dbname=$dbname;charset=utf8mb4", $username, $password, [
        PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
        PDO::ATTR_PERSISTENT => true
    ]);

    // Get the start and end dates from query parameters
    /*$startTime = '00:00:00';*/
    $startDate = $_GET['startDate'] . ' ' . $_GET['startTime']; // Append time to start date
    $endDate = $_GET['endDate'] . ' 23:59:59'; // Append time to end date
    // Convert the local times to UTC (assuming input is in 'America/New_York')
    $startDateTime = new DateTime($startDate, new DateTimeZone('America/New_York'));
    $endDateTime = new DateTime($endDate, new DateTimeZone('America/New_York'));

    // Convert to UTC
    $startDateTime->setTimezone(new DateTimeZone('UTC'));
    $endDateTime->setTimezone(new DateTimeZone('UTC'));

    // Format the dates for the SQL query
    $startDateUTC = $startDateTime->format('Y-m-d H:i:s');
    $endDateUTC = $endDateTime->format('Y-m-d H:i:s');

    // Calculate the interval based on the date range
    $startDateTime = new DateTime($startDate);
    $endDateTime = new DateTime($endDate);
    $interval = $startDateTime->diff($endDateTime);
    $groupInterval = 60;  //Data shown for all 60 sec of a minute
    /*// Check if the interval is exactly one day
    if ($interval->days >= 0 && $interval->days < 3) {
    // Use 1-minute intervals for a single day
    $groupInterval = '1 MINUTE';
    }elseif ($interval->days <= 7) {
        // Use 5-minute intervals for a week range
        $groupInterval = '1 MINUTE';
    } elseif ($interval->days <= 30) {
        // Use 10-minute intervals for a month range
        $groupInterval = '1 MINUTE';
        $groupInterval = '10 MINUTE';
    }*/

    //Change table name "renewable_data" after 'FROM' based on your table name
    $stmt = $pdo->prepare("
        SELECT
            DATE_FORMAT(DATE_SUB(date_time, INTERVAL FLOOR(SECOND(date_time) / :interval) * :interval SECOND), '%Y-%m-%d %H:%i:%s') AS interval_time,
            solar_percentage,
            wind_percentage,
            hydro_percentage,
            battery_percentage
        FROM historical_data
        WHERE date_time BETWEEN :startDate AND :endDate
        GROUP BY interval_time
        ORDER BY interval_time ASC
    ");

    $stmt->bindValue(':interval', $groupInterval, PDO::PARAM_INT);
    $stmt->bindParam(':startDate', $startDateUTC);
    $stmt->bindParam(':endDate', $endDateUTC);
    $stmt->execute();

    // Fetch results
    $data = [
        'solar' => [],
        'wind' => [],
        'hydro' => [],
        'battery' => [],
        'interval_times' => []
    ];

    $results = $stmt->fetchAll(PDO::FETCH_ASSOC);

    if (empty($results)) {
        echo json_encode(['error' => 'No data found for the given date range']);
        exit;
    }

    foreach ($results as $row) {
      // Adjust the datetime to the local time zone
        $dateTime = new DateTime($row['interval_time'], new DateTimeZone('UTC')); // Assuming the data is stored in UTC
        $dateTime->setTimezone(new DateTimeZone('America/New_York')); // Convert to local timezone
        $localDateTime = $dateTime->format('Y-m-d H:i:s'); // Format as needed

        $data['interval_times'][] = $localDateTime;
        $data['solar'][] = (float)$row['solar_percentage'];
        $data['wind'][] = (float)$row['wind_percentage'];
        $data['hydro'][] = (float)$row['hydro_percentage'];
        $data['battery'][] = (float)$row['battery_percentage'];
    }

    echo json_encode($data, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
} catch (PDOException $e) {
    echo json_encode(['error' => "Connection failed: " . $e->getMessage()]);
}
?>

File: current/python-db-scripts/fetchData.py
from fastapi import FastAPI, Query
from fastapi.middleware.cors import CORSMiddleware
import pymysql
import pymysql.cursors
from datetime import datetime
import pytz
from config import servername, username, password, dbname

app = FastAPI()

app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_methods=["*"],
    allow_headers=["*"],
)

def get_connection():
    return pymysql.connect(
        host=servername,
        user=username,
        password=password,
        database=dbname,
        charset='utf8mb4',
        cursorclass=pymysql.cursors.DictCursor
    )

@app.get("/data")
def get_data(
    startDate: str = Query(...),
    startTime: str = Query(...),
    endDate: str = Query(...),
):
    eastern = pytz.timezone("America/New_York")
    utc = pytz.utc

    start_local = eastern.localize(datetime.strptime(f"{startDate} {startTime}", "%Y-%m-%d %H:%M:%S"))
    end_local = eastern.localize(datetime.strptime(f"{endDate} 23:59:59", "%Y-%m-%d %H:%M:%S"))

    start_utc = start_local.astimezone(utc).strftime("%Y-%m-%d %H:%M:%S")
    end_utc = end_local.astimezone(utc).strftime("%Y-%m-%d %H:%M:%S")

    group_interval = 300

    sql = """
    SELECT
        interval_time,
        ROUND(AVG(solar_percentage), 2)       AS solar_percentage,
        ROUND(AVG(wind_percentage), 2)        AS wind_percentage,
        ROUND(AVG(hydro_percentage), 2)       AS hydro_percentage,
        ROUND(AVG(battery_percentage), 2)     AS battery_percentage,
        ROUND(AVG(solar_fixed_percentage), 2) AS solar_fixed_percentage,
        ROUND(AVG(solar_360_percentage), 2)   AS solar_360_percentage
    FROM (
        SELECT
            date_time,
            solar_percentage, wind_percentage, hydro_percentage,
            battery_percentage, solar_fixed_percentage, solar_360_percentage,
            FROM_UNIXTIME(FLOOR(UNIX_TIMESTAMP(date_time) / %s) * %s) AS interval_time
        FROM historical_data
        WHERE date_time >= %s AND date_time <= %s
    ) t
    GROUP BY interval_time
    ORDER BY interval_time ASC
    """

    connection = get_connection()
    try:
        with connection.cursor() as cursor:
            cursor.execute(sql, (group_interval, group_interval, start_utc, end_utc))
            results = cursor.fetchall()
    finally:
        connection.close()

    if not results:
        return {"error": "No data found for the given date range"}

    data = {
        "solar": [], "wind": [], "hydro": [], "battery": [],
        "solarFixed": [], "solar360": [], "interval_times": []
    }

    for row in results:
        dt_utc = row["interval_time"]
        if isinstance(dt_utc, str):
            dt_utc = datetime.strptime(dt_utc, "%Y-%m-%d %H:%M:%S")
        dt_local = utc.localize(dt_utc).astimezone(eastern).strftime("%Y-%m-%d %H:%M:%S")

        data["interval_times"].append(dt_local)
        data["solar"].append(float(row["solar_percentage"] or 0))
        data["wind"].append(float(row["wind_percentage"] or 0))
        data["hydro"].append(float(row["hydro_percentage"] or 0))
        data["battery"].append(float(row["battery_percentage"] or 0))
        data["solarFixed"].append(float(row["solar_fixed_percentage"] or 0))
        data["solar360"].append(float(row["solar_360_percentage"] or 0))

    return data